### PR TITLE
feat(in-page-navigation): add new component package

### PR DESCRIPTION
**Pull request by gloriliale**
# In this PR
This PR creates the new package In Page Navigation.

~~Right now, this is a draft with one story for preliminary design review.~~
This PR is ready for review!

- [x] Stories
- [x] Customization
- [x] Tests
- [x] Docs page 

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [ ] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    },
+  };
+};
+
+export interface InPageNavigationItemProps extends React.AnchorHTMLAttributes<HTMLAnchorElement> {
+  currentPage?: boolean;
+  children: NonNullable<React.ReactNode>;
+  href: string;
+  ref?: any;
+  rel?: string; // do we need this? I don't understand how we're using it in Anchor
```
Comment (gloriliale) [inline packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx:50]: I added this prop bc it's in our Anchor component, but when I looked at Anchor, I didn't truly understand what we did with it there. The main question is, do we need this prop in InPageNavigationItem?

File: packages/paste-core/components/in-page-navigation/__tests__/index.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      <InPageNavigation aria-label="my-nav">
+        <InPageNavigationItem href="#">page 1</InPageNavigationItem>
+        <InPageNavigationItem href="#">page 2</InPageNavigationItem>
+      </InPageNavigation>
+    );
+
+    expect(getByRole('navigation').getAttribute('aria-label')).toBe('my-nav');
+  });
+
+  it('should render a list with list items', () => {
```
Comment (shleewhite) [inline packages/paste-core/components/in-page-navigation/__tests__/index.spec.tsx:19]: Suggestion(non-blocking): could this render test and the "should render anchors" test be combined? that would reduce renders

File: packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx (hunk 1 above)
Comment (shleewhite) [reply packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx:50]: We probably should, its a HTML attribute: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel#:~:text=The%20rel%20attribute%20defines%20the,which%20the%20attribute%20is%20found.

Usually I see people set it to noreferrer.. 

However, I think it should be included in `React.AnchorHTMLAttributes`- not sure why it would need to be specified here too.

File: packages/paste-core/components/in-page-navigation/stories/index.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+import {InPageNavigation, InPageNavigationItem} from '../src';
+
+// eslint-disable-next-line import/no-default-export
+export default {
+  title: 'Components/In Page Navigation',
+  component: InPageNavigation,
+};
+
+export const Default: Story = () => {
```
Comment (shleewhite) [inline packages/paste-core/components/in-page-navigation/stories/index.stories.tsx:18]: Suggestion (non-blocking): do we need to have a comparison in the story?

File: packages/paste-website/src/component-examples/InPageNavigationExamples.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
+export const basicExample = `
+const BasicInPageNavigation = () => {
+
+  return (
+    <InPageNavigation aria-label="my-nav">
```
Comment (shleewhite) [inline packages/paste-website/src/component-examples/InPageNavigationExamples.ts:5]: Issue(blocking): aria label has to be a human readable string. So it could be "Primary navigation" or something like that. Same with the full width example.

**Review by kelseycavitt — COMMENTED** [review]
@gloriliale Similar to what we did for the radio button group component ([tabs](https://paste.twilio.design/components/radio-button-group/#radio-button-groups-vs-tabs), [radio group](https://paste.twilio.design/components/radio-button-group/#button-group-vs-radio-button-group)) , could we have a lil paragraph/sentence about the difference between Tabs and In Page Navigation?

File: packages/paste-website/src/component-examples/InPageNavigationExamples.ts (hunk 4 above)
Comment (nkrantz) [reply packages/paste-website/src/component-examples/InPageNavigationExamples.ts:5]: the stories should also be updated to human readable strings

File: packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        ref={ref}
+        element={element}
+        display="flex"
+        flexBasis={variant === 'fullWidth' ? '50%' : undefined}
+        flexGrow={variant === 'fullWidth' ? 1 : undefined}
+        flexShrink={variant === 'fullWidth' ? 1 : undefined}
+      >
+        <Box
+          {...safelySpreadBoxProps(props)}
+          {...getItemStyles(currentPage, variant)}
```
Comment (TheSisb) [inline packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx:70]: thought: looks like you're applying the fullWidth styles to this inner element and the outer `li`. Probably only need it on the outer `li`.

File: packages/paste-core/components/in-page-navigation/stories/index.stories.tsx (hunk 3 above)
Comment (TheSisb) [reply packages/paste-core/components/in-page-navigation/stories/index.stories.tsx:18]: +1 to removing the comparison

File: packages/paste-core/components/in-page-navigation/stories/index.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  return (
+    <>
+      <Heading as="h1" variant="heading30">
+        Comparison to Tabs style
+      </Heading>
+      <Stack orientation="vertical" spacing="space120">
+        <Box>
+          This is what <strong>In Page Navigation</strong> looks like (only for visual- clicking should not lead
+          anywhere)
+          <InPageNavigation aria-label="myNav">
```
Comment (TheSisb) [inline packages/paste-core/components/in-page-navigation/stories/index.stories.tsx:28]: todo (blocking): let's try and look on the web or in product for a real example with a human readable label. Let's use at least 2 different realistic examples on our storybook and website docs.

File: packages/paste-core/components/in-page-navigation/build.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+const {build} = require('../../../../tools/build/esbuild');
```
[the remark this answers is not in the document]
Comment (TheSisb) [reply packages/paste-core/components/in-page-navigation/build.js:1]: In the future I plan to refactor it into a private `scripts` package, but haven't done so yet

File: packages/paste-core/components/in-page-navigation/__tests__/index.spec.tsx (hunk 2 above)
Comment (gloriliale) [reply packages/paste-core/components/in-page-navigation/__tests__/index.spec.tsx:19]: yes yes, good idea

File: packages/paste-core/components/in-page-navigation/stories/index.stories.tsx (hunk 3 above)
Comment (gloriliale) [reply packages/paste-core/components/in-page-navigation/stories/index.stories.tsx:18]: I left the comparisons in to make it easier for @twilio-labs/design-systems-pd to review. Maybe this is something we can remove after they've approved it?

File: packages/paste-core/components/in-page-navigation/__tests__/index.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        <InPageNavigationItem currentPage href="#">
+          page 2
+        </InPageNavigationItem>
+      </InPageNavigation>
+    );
+
+    expect(getByText('page 2').getAttribute('aria-current')).toBe('page');
+  });
+
+  it('should pass props given to InPageNavigationItem onto its Anchor child', () => {
```
[the remark this answers is not in the document]
Comment (gloriliale) [reply packages/paste-core/components/in-page-navigation/__tests__/index.spec.tsx:55]: great callout, it's the `a` element. Will update accordingly!

File: packages/paste-core/components/in-page-navigation/src/InPageNavigation.tsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+import type {BoxProps} from '@twilio-paste/box';
+import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
+
+import {InPageNavigationContext} from './InPageNavigationContext';
+
+export interface InPageNavigationProps extends Omit<React.ComponentPropsWithRef<'div'>, 'children'> {
+  children: React.ReactNode;
+  element?: BoxProps['element'];
+  'aria-label': string;
+  variant?: 'fullWidth';
```
[the remark this answers is not in the document]
Comment (gloriliale) [reply packages/paste-core/components/in-page-navigation/src/InPageNavigation.tsx:12]: ooo good call, adding this now!

File: packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx (hunk 5 above)
Comment (gloriliale) [reply packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx:70]: I'm not sure exactly why, but it does need to be in both places for the inner text to be centered. Checked this out w/ @SiTaggart yesterday. Open to suggestions if there's anything that makes more logical sense though.

File: packages/paste-core/components/in-page-navigation/__tests__/index.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    const {getByText} = render(
+      <InPageNavigation aria-label="my-nav">
+        <InPageNavigationItem href="#">page 1</InPageNavigationItem>
+        <InPageNavigationItem currentPage href="#">
+          page 2
+        </InPageNavigationItem>
+      </InPageNavigation>
+    );
+
+    expect(getByText('page 2').getAttribute('aria-current')).toBe('page');
```
[the remark this answers is not in the document]
Comment (gloriliale) [reply packages/paste-core/components/in-page-navigation/__tests__/index.spec.tsx:52]: oh i definitely like this better, updating now!

File: packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx [hunk 11]
```diff
@@ -0,0 +1,95 @@
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+import type {BoxStyleProps, BoxProps} from '@twilio-paste/box';
+import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
+
+import {InPageNavigationContext} from './InPageNavigationContext';
+
+const getItemStyles = (currentPage: boolean, variant?: string): BoxStyleProps => {
```
[the remark this answers is not in the document]
Comment (gloriliale) [reply packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx:8]: @TheSisb I did this but I'd like to take a look at it together if you've got some time tomorrow (Friday).

File: packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx (hunk 1 above)
Comment (gloriliale) [reply packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx]: update: 
- `rel` and `target` are included in `React.AnchorHTMLAttributes` as @shleewhite pointed out above, so it has been removed from the props
- `href`, though also included in `React.AnchorHTMLAttributes`, is optional there while we want it to be required here, so it is declared here to be required.
- `ref` is not necessary to add to our props, so it has been removed. @TheSisb and I tested this out by removing this line and adding a ref to a story. It worked.

File: packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (gloriliale) [reply packages/paste-core/components/in-page-navigation/src/InPageNavigationItem.tsx]: update: refactored this even better, woohoo! thx @TheSisb!

File: packages/paste-core/components/in-page-navigation/stories/index.stories.tsx (hunk 3 above)
Comment (kelseycavitt) [reply packages/paste-core/components/in-page-navigation/stories/index.stories.tsx:15]: I think having it initially was good, but post review, let's remove it!

**Review by kelseycavitt — COMMENTED** [review]
@gloriliale 
Something that didn't occur to me to try with my first review was to stress test the # of items within the component, SO sorry about that! I did end up finding a little wonkiness when there's a large number or longer names of an item. I think the longer names is very edge case, but the number of items for smaller screen widths is a use case worth addressing I think.

1. Here's a [**Loom**](https://www.loom.com/share/7724fb2867a449179312cfab0b3a38f4) that shows the wonkiness in the In Page Nav component. I tried this out on the docs site page as well.

2. Here's another [**Loom**](https://www.loom.com/share/6feba25fbfe644749163c70930642760) that shows what we do for tabs

I think the solution here would be to actually not truncate (like tabs does) and allow a user to horizontally scroll within the component. Both things aren't super ideal from a design perspective (specifically if you're on a web browser, vs. a mobile web browser). Let me know what you think!

**Review by TheSisb — DISMISSED** [review]

**Review by loreina — DISMISSED** [review]

**Review by nkrantz — DISMISSED** [review]

**Review by TheSisb — APPROVED** [review]

**Review by nkrantz — APPROVED** [review]

**Review by loreina — APPROVED** [review]

**Review by shleewhite — APPROVED** [review]
